### PR TITLE
Expand Microsoft.ContainerService/managedClusters/agentPool properties object

### PR DESCRIPTION
**Pull request by lsnoddy**
# Change

Added all available properties as parameters for the agentPool module instead of relying on a single agentPoolProperties object.

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: .github/workflows/platform.dependencies.yml [hunk 1]
```diff
@@ -915,4 +915,4 @@ jobs:
           resourceGroupName: '${{ env.resourceGroupName }}'
           subscriptionId: '${{ secrets.ARM_SUBSCRIPTION_ID }}'
           managementGroupId: '${{ secrets.ARM_MGMTGROUP_ID }}'
-          removeDeployment: '${{ env.removeDeployment }}'
+          removeDeployment: '${{ env.removeDeployment }}'
```
Comment (MrMCake) [inline .github/workflows/platform.dependencies.yml:918]: Was that intended?

File: README.md [hunk 2]
```diff
@@ -171,4 +171,4 @@ Any use of third-party trademarks or logos are subject to those third-party's po
 
 <!-- Docs -->
 [MicrosoftAzureDocs]: <https://docs.microsoft.com/en-us/azure/>
-[PowerShellDocs]: <https://docs.microsoft.com/en-us/powershell/>
+[PowerShellDocs]: <https://docs.microsoft.com/en-us/powershell/>
```
Comment (MrMCake) [inline README.md:174]: Was that intended?

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep [hunk 3]
```diff
@@ -5,17 +5,503 @@ param managedClusterName string
...
+@description('Optional. The node labels to be persisted across all nodes in agent pool.')
+param nodeLabels object = {}
+
+@description('Optional. ResourceId of the node PublicIPPrefix')
+param nodePublicIpPrefixId string = ''
+
+@description('Optional. The taints added to new nodes during node pool create and scale. For example, key=value:NoSchedule.	')
+param nodeTaints array = []
+
+@description('Optional. 	As a best practice, you should upgrade all node pools in an AKS cluster to the same Kubernetes version. The node pool version must have the same major version as the control plane. The node pool minor version must be within two minor versions of the control plane version. The node pool version cannot be greater than the control plane version. For more information see upgrading a node pool (https://docs.microsoft.com/en-us/azure/aks/use-multiple-node-pools#upgrade-a-node-pool).')
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:322]: Remove redundant spaces
```suggestion
@description('Optional. As a best practice, you should upgrade all node pools in an AKS cluster to the same Kubernetes version. The node pool version must have the same major version as the control plane. The node pool minor version must be within two minor versions of the control plane version. The node pool version cannot be greater than the control plane version. For more information see upgrading a node pool (https://docs.microsoft.com/en-us/azure/aks/use-multiple-node-pools#upgrade-a-node-pool).')
```

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep [hunk 4]
```diff
@@ -5,17 +5,503 @@ param managedClusterName string
...
+@description('Optional. VM size. VM size availability varies by region. If a node contains insufficient compute resources (memory, cpu, etc) pods might fail to run correctly. For more details on restricted VM sizes, see: /azure/aks/quotas-skus-regions')
+param vmSize string = 'Standard_D2s_v3'
+
+@description('Optional. Node Subnet IDIf this is not specified, a VNET and subnet will be generated and used. If no podSubnetID is specified, this applies to nodes and pods, otherwise it applies to just nodes. This is of the form: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/virtualNetworks/{virtualNetworkName}/subnets/{subnetName}	')
+param vnetSubnetId string = ''
+
+@description('Optional. Determines the type of workload a node can run.')
+param workloadRuntime string = ''
 
 resource managedCluster 'Microsoft.ContainerService/managedClusters@2021-08-01' existing = {
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:400]: I'd suggest to move this reference down to the agent pool resource

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep [hunk 5]
```diff
@@ -5,17 +5,503 @@ param managedClusterName string
...
+    enableAutoScaling: enableAutoScaling
+    enableEncryptionAtHost: enableEncryptionAtHost
+    enableFIPS: enableFIPS
+    enableNodePublicIP: enableNodePublicIP
+    enableUltraSSD: enableUltraSSD
+    gpuInstanceProfile: !empty(gpuInstanceProfile) ? gpuInstanceProfile : null
+    kubeletConfig: kubeletConfig
+    kubeletDiskType: kubeletDiskType
+    linuxOSConfig: linuxOSConfig
+    maxCount: !(maxCount == -1) ? maxCount : null
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:480]:
```suggestion
    maxCount: maxCount != -1 ? maxCount : null
```

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep [hunk 6]
```diff
@@ -5,17 +5,503 @@ param managedClusterName string
...
+    maxPods: !(maxPods == -1) ? maxPods : null
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:481]:
```suggestion
    maxPods: maxPods != -1 ? maxPods : null
```

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep [hunk 7]
```diff
@@ -5,17 +5,503 @@ param managedClusterName string
...
+    minCount: !(minCount == -1) ? minCount : null
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:482]:
```suggestion
    minCount: minCount != -1 ? minCount : null
```

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep [hunk 8]
```diff
@@ -5,17 +5,503 @@ param managedClusterName string
...
+    mode: !empty(mode) ? mode : null
+    nodeLabels: nodeLabels
+    nodePublicIPPrefixID: !empty(nodePublicIpPrefixId) ? nodePublicIpPrefixId : null
+    nodeTaints: nodeTaints
+    orchestratorVersion: orchestratorVersion
+    osDiskSizeGB: !(osDiskSizeGB == -1) ? osDiskSizeGB : null
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:488]:
```suggestion
    osDiskSizeGB: osDiskSizeGB != -1 ? osDiskSizeGB : null
```

File: arm/Microsoft.ContainerService/managedClusters/agentPools/readme.md [hunk 9]
```diff
@@ -6,23 +6,98 @@ This module deploys an Agent Pool for a Container Service Managed Cluster
...
+| `type` | string |  |  | Optional. The type of Agent Pool. |
+| `vmMaxMapCount` | int | `65530` |  | Optional. Sysctl setting vm.max_map_count. |
+| `vmSize` | string | `Standard_D2s_v3` |  | Optional. VM size. VM size availability varies by region. If nodes contains insufficient compute resources (memory, cpu, etc) pods might fail to run correctly. For more details on restricted VM sizes, see: /azure/aks/quotas-skus-regions |
+| `vmSwappiness` | int |  |  | Optional. Sysctl setting vm.swappiness. |
+| `vmVfsCachePressure` | int |  |  | Optional. Sysctl setting vm.vfs_cache_pressure. |
+| `vnetSubnetId` | string |  |  | Optional. Node-Subnet ID. If this is not specified, a VNET and subnet will be generated and used. If no podSubnetID is specified, this applies to nodes and pods, otherwise it applies to just nodes. This is of the form: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/virtualNetworks/{virtualNetworkName}/subnets/{subnetName} |
+| `workloadRuntime` | string |  |  | Optional. Determines the type of workload nodes can run. |
 
-### Parameter Usage: `agentPoolProperties`
+### Parameter Usage: `agentPools`
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/readme.md:94]: We don't need this example as it is essentially just the resource itself

File: arm/Microsoft.ContainerService/managedClusters/readme.md [hunk 10]
```diff
@@ -183,50 +183,47 @@ For available properties check <https://docs.microsoft.com/en-us/azure/templates
 
 ### Parameter Usage: `agentPools`
 
+Provide values for additional agent pools as needed.
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/readme.md:186]: Can be removed as the parameters table automatically links to the agent pool readme

File: arm/README.md [hunk 11]
```diff
@@ -86,4 +86,4 @@ In this section you can find useful information regarding the Modules that are c
 | [API Connections](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/connections) | `MS.Web` | [connections](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/connections) |
 | [App Service Environments](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/hostingEnvironments) |  | [hostingEnvironments](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/hostingEnvironments) |
 | [App Service Plans](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/serverfarms) |  | [serverfarms](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/serverfarms) |
-| [Web/Function Apps](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/sites) |  | [sites](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/sites) |
+| [Web/Function Apps](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/sites) |  | [sites](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/sites) |
```
Comment (MrMCake) [inline arm/README.md:89]: Please undo as the readme will be autogenerated

File: .github/workflows/platform.dependencies.yml (hunk 1 above)
Comment (lsnoddy) [reply .github/workflows/platform.dependencies.yml]: Unintended. Restored the file to remove any changes.

File: README.md (hunk 2 above)
Comment (lsnoddy) [reply README.md]: Unintended, restored the file to remove any changes.

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep (hunk 3 above)
Comment (lsnoddy) [reply arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:322]: Removed extra spaces.

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep (hunk 5 above)
Comment (lsnoddy) [reply arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:480]: Changed to not equal !=

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep (hunk 6 above)
Comment (lsnoddy) [reply arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:481]: Changed to not equal !=

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep (hunk 7 above)
Comment (lsnoddy) [reply arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:482]: Changed to not equal !=

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep (hunk 8 above)
Comment (lsnoddy) [reply arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:488]: Changed to not equal !=

File: arm/Microsoft.ContainerService/managedClusters/agentPools/readme.md (hunk 9 above)
Comment (lsnoddy) [reply arm/Microsoft.ContainerService/managedClusters/agentPools/readme.md:94]: Removed the example

File: arm/Microsoft.ContainerService/managedClusters/readme.md (hunk 10 above)
Comment (lsnoddy) [reply arm/Microsoft.ContainerService/managedClusters/readme.md:186]: Removed section

File: arm/README.md [hunk 12]
```diff
@@ -57,7 +57,6 @@ In this section you can find useful information regarding the Modules that are c
 | [Virtual Network Gateway Connections](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Network/connections) |  | [connections](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Network/connections) |
 | [DDoS Protection Plans](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Network/ddosProtectionPlans) |  | [ddosProtectionPlans](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Network/ddosProtectionPlans) |
 | [ExpressRoute Circuits](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Network/expressRouteCircuits) |  | [expressRouteCircuits](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Network/expressRouteCircuits) |
-| [Network Firewall Policies](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Network/firewallPolicies) |  | [firewallPolicies](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Network/firewallPolicies) |
```
Comment (MrMCake) [inline arm/README.md:60]: That's strange. The module does exit in your branch

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep (hunk 3 above)
Comment (MrMCake) [reply arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:322]: are you sure?

File: arm/README.md (hunk 12 above)
Comment (lsnoddy) [reply arm/README.md]: Restored this file, should resolve discrepancy

File: arm/README.md (hunk 11 above)
Comment (lsnoddy) [reply arm/README.md]: Reverted changes to this file

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep (hunk 4 above)
Comment (lsnoddy) [reply arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep]: Moved reference down.

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep (hunk 3 above)
Comment (lsnoddy) [reply arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep]: Was in the middle up making changes, latest code has been uploaded.

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep [hunk 13]
```diff
@@ -5,17 +5,503 @@ param managedClusterName string
...
+    enableAutoScaling: enableAutoScaling
+    enableEncryptionAtHost: enableEncryptionAtHost
+    enableFIPS: enableFIPS
+    enableNodePublicIP: enableNodePublicIP
+    enableUltraSSD: enableUltraSSD
+    gpuInstanceProfile: !empty(gpuInstanceProfile) ? gpuInstanceProfile : null
+    kubeletConfig: kubeletConfig
+    kubeletDiskType: kubeletDiskType
+    linuxOSConfig: linuxOSConfig
+    maxCount: (maxCount != -1) ? maxCount : null
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:480]:
```suggestion
    maxCount: maxCount != -1 ? maxCount : null
```

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep [hunk 14]
```diff
@@ -5,17 +5,503 @@ param managedClusterName string
...
+    maxPods: (maxPods != -1) ? maxPods : null
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:481]:
```suggestion
    maxPods: maxPods != -1 ? maxPods : null
```

File: arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep [hunk 15]
```diff
@@ -5,17 +5,503 @@ param managedClusterName string
...
+    minCount: (minCount != -1) ? minCount : null
```
Comment (MrMCake) [inline arm/Microsoft.ContainerService/managedClusters/agentPools/deploy.bicep:482]:
```suggestion
    minCount: minCount != -1 ? minCount : null
```

**Review by MrMCake — APPROVED** [review]

**Pull request merged**